File: server/utils/gencsv.py
import csv, sys

PREFIX = "CY"
CSV_OUTPUT_FILE = "list.csv"
# maximum product id possible = 36^6 = 2176782336; i.e. [0, 2176782336)
MAX_PRODUCT_ID = 2176782336
BASE = 36

def validate(start, count):
    end = start + count
    # ensure count and start is +ve
    assert (start >= 0 and count > 0)
    assert (start < MAX_PRODUCT_ID)
    assert (end <= MAX_PRODUCT_ID)

# Function to convert a number to the extended base-36 string
def convert_number_to_base36(number):
    if 0 <= number < BASE**6:
        base36_chars = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'
        base36_str = ''
        for _ in range(6):
            number, remainder = divmod(number, BASE)
            base36_str = base36_chars[remainder] + base36_str
        return PREFIX + base36_str
    else:
        raise ValueError("Number must be between 0 and 36^6-1 for this conversion.")


if __name__ == "__main__":
    # called from terminal

    # numbers in the range -2147483648 through 2147483647
    start = int(sys.argv[1])
    count = int(sys.argv[2])
    end = start + count
    PREFIX = sys.argv[3]

    validate(start, count)

    # generate the list [start, end)
    # write for the tracker sheet entry
    validate(start, count)
    with open('tracker.csv', 'w') as new_file:
        write=csv.writer(new_file, delimiter='\n')
        [write.writerow([convert_number_to_base36(id)]) for id in range(start, end)]



# >>> import execution_methods
# This is my file to test Python's execution methods.
# The variable __name__ tells me which context this file is running in.
# The value of __name__ is: 'execution_methods'

File: server/static/js/gencsv.js
var PREFIX = "CY";
const BASE = 36;
const MAX_PRODUCT_ID = 2176782336;

function validate(start, count) {
    const end = start + count;
    // ensure count and start are positive and within bounds
    if (isNaN(start) || isNaN(count) || !(start >= 0 && count > 0 && start < MAX_PRODUCT_ID && end <= MAX_PRODUCT_ID)) {
        throw new Error("Invalid start or count values");
    }
}

// Function to convert a number to the extended base-36 string
function convertNumberToBase36(number) {
    if (0 <= number < BASE**6) {
        const base36Chars = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';
        let base36Str = '';
        for (let i = 0; i < 6; i++) {
            const [quotient, remainder] = [Math.floor(number / BASE), number % BASE];
            number = quotient;
            base36Str = base36Chars[remainder] + base36Str;
        }
        return PREFIX + base36Str;
    } else {
        throw new Error("Number must be between 0 and 36^6-1 for this conversion.");
    }
}

function generate_skus(start, count, prefix) {
    PREFIX = prefix
    validate(start, count);

    // Generate the list [start, end)
    const result = Array.from({ length: count }, (_, id) => convertNumberToBase36(start + id));
    console.table(result);
    return result.join('\n');
}
